### ISSUE TITLE
Java backend doesn't know how to pull a term out of a strict position when putting a K sequence in that position

### ISSUE THREAD

**Issue by liyili2** [kframework/k-legacy#745]
When I have this set of rules:

```
module TEST

     syntax KItem ::= maybeSymVal(K, K) [strict]
                | "checkResult" | "start" 

     syntax KResult ::= "noResult" | "symVal" | TypedValue

     rule R:KResult ~> checkResult => R
     rule checkResult => noResult

     syntax TypedValue ::= tv(K, Int) 


     syntax PureStatement ::= "BlockStatement" "(" K ")"
          [klabel('BlockStatement)]

     rule start => maybeSymVal(BlockStatement(1) ~> checkResult,BlockStatement(0) ~> checkResult)

     rule BlockStatement(1) => tv(1,1)
     rule BlockStatement(0) => .K

     rule maybeSymVal(_:KResult, noResult) => .K
     rule maybeSymVal(tv(_, T:Int), tv(_, T:Int)) => tv(symVal, T)

endmodule

```

If I run the program "start" in Java Backend, it will give me the following result:

```
<k>
    maybeSymVal ( BlockStatement ( 1 ) ~> checkResult , BlockStatement ( 0 )
       ~> checkResult )
</k>

```

However, if I run the program in Maude Backend, it will give me the following result:

```
<k>
    .K
</k>
```

I also ask Java Backend to print out the sequence of rules it actually executes. It seems that Java Backend never applies the strict pulling out rule on the two strict positions in the constructor "maybeSymVal". Actually, I don't know how to re-factor this set of rules to make it work in Java Backend. I wonder if someone can help me either solve the problem or tell me how to re-factor it.

**Comment by yilongli** [comment]
@andreistefanescu you may want to take a look at this. In the following commit, you added the comment saying that `/* the java rewrite engine only supports heating/cooling on KItem */`:
https://github.com/kframework/k/commit/cef7b36c371f4e8d11ac38c69e9ba88e0239f74e

How should I proceed?

**Comment by liyili2** [comment]
```
module TEST

     syntax KItem ::= maybeSymVal(K, K) [strict]
                | checkResult(K) [strict] | "start" 

     syntax KResult ::= "noResult" | "symVal" | TypedValue

     rule checkResult(R:KResult)  => R
     rule checkResult(HOLE) => noResult

     syntax TypedValue ::= tv(K, Int) 


     syntax PureStatement ::= "BlockStatement" "(" K ")"
          [klabel('BlockStatement)]

     rule start => maybeSymVal(checkResult(BlockStatement(1)),checkResult(BlockStatement(0)))

     rule BlockStatement(1) => tv(1,1)
     rule BlockStatement(0) => .K

     rule maybeSymVal(_:KResult, noResult) => .K
     rule maybeSymVal(tv(_, T:Int), tv(_, T:Int)) => tv(symVal, T)

endmodule
```

I tried this in Java backend, but it does not work. I really do not want to change every rules which will be the content of the constructor "maybeSymVal". Can anyone suggest a easy way for me to re-design  the semantic to solve the problem.

**Comment by grosu** [comment]
One thing which caught my eye here is that

```
syntax PureStatement ::= "BlockStatement" "(" K ")"  [klabel('BlockStatement)]
```

should be completely equivalent to

```
syntax PureStatement ::= BlockStatement(K)
```

So what's the problem with this example?  What exactly is that it does not work?  What did you try and didn't work?

Grigore

**Comment by liyili2** [comment]
Yilong has suggested the problem actually. The problem is that current Java Backend cannot have a K sequence term instead of KItem term in heating/cooling rules. The maybeSymVal constructor is defined as strict, and when I specify a rule and put the term "BlockStatement(1) ~> checkResult" in the first block of the constructor "maybeSymVal", Java backend doesn't know how to pull a term out according to this rule, because the term will be a K Sequence term instead of a KItem.  

That is why Yilong indicated in his response.

**Comment by grosu** [comment]
Liyi, write good English!  Do something about it and focus.  It is very annoying to see how you keep making the same mistakes over and over again.  You will eventually have to write papers and I swear that I am NOT going to write your papers, and nobody from our group will!

The fact that you can only heat on KItems does not seem to apply in your second example, because you only have KItems.  So what is the problem here?

**Comment by liyili2** [comment]
The second one is not a example. It is the fact that I tried to re-factor the set of rules by rewriting a term with a HOLE directly, but I find that Java Backend doesn't allow me to do that, and I think it is very bad feeling for me to rewrite a term with a HOLE.

**Comment by yilongli** [comment]
Regarding your second example, the result is the following:

```
<k>
    checkResult ( HOLE ) ~> maybeSymVal ( tv ( 1 , 1 ) , HOLE )
</k>
```

This rule should be able to apply:

```
rule checkResult(HOLE) => noResult
```

So it's a bug. I will fix it ASAP. If you could paste the above result and point out which rule you think should apply in the first place, it would save both of us time.

**Comment by yilongli** [comment]
I said in the previous reply that failing to apply `rule checkResult(HOLE) => noResult` was a bug in the Java backend. However, after some investigation, I have to take it back. In the final configuration, what gets stuck at top of k cell is a freezer, while in the rule `checkResult(HOLE)` is a normal KItem. Therefore, the behavior of the Java rewrite engine is expected. In order for this rule to be applied, `checkResult(HOLE)` should be constructed as a freezer before it's passed to the Java rewrite engine.

**Comment by liyili2** [comment]
I report the bug yesterday. The second example is just a way to fix the problem quickly. However, I don't want to use the second example, since it is very odd to look at the rule based on a person who writes programming language rules. The rule:

```
checkResult ( HOLE ) => noResult
```

is very odd. It is very hard and shy for me to write down this rule in a paper in any semantic rule system. I would rather someone to help me allow heating/cooling rules on a K sequence term.

**Comment by liyili2** [comment]
@yilongli , would you mind to tell me how to construct a freezer for the term?

**Comment by yilongli** [comment]
I am not sure if there is a way to specify a freezer in the K definition. Maybe someone more experienced in K can help you.

**Comment by yilongli** [comment]
In the second example, you pasted a definition and said "I tried this in Java backend, but it does not work." However, as Grigore pointed out, you didn't say what didn't work, what you tried, and the expected result. Writing clear bug report benefits both of us.

**Comment by yilongli** [comment]
To avoid getting blocked by this issue when there is still decision to be made regarding whether we should allow heating on K, you can change this line:
https://github.com/kframework/k/blob/bcbfff724c1271995def1c48503126b2e2fcf48a/src/javasources/KTool/src/org/kframework/compile/transformers/ContextsToHeating.java#L60
to `v = Variable.getFreshVar(KSorts.K);`.
This will allow heating on K.

**Comment by liyili2** [comment]
@yilongli , Do you have the design of heating on K in Java Backend yet.

**Comment by yilongli** [comment]
I don't understand your question.

**Comment by liyili2** [comment]
I mean, even if I change the line you pointed to be `v = Variable.getFreshVar(KSorts.K);`, if you don't have the support for heating on K in your rewriting engine, it will not work still. I wonder if the Java Backend will have heating on K work correctly if I change the line you pointed out to be  `v = Variable.getFreshVar(KSorts.K);`.

**Comment by yilongli** [comment]
If it wouldn't work, why would I suggest you to change this? Yes, there may be bugs, but I believe that is what needs to be changed in order to support heating on K. Of course, there may be implications on Andrei's prover which I am not aware of. Why not give it a try yourself?

**Comment by liyili2** [comment]
ok

**Comment by grosu** [comment]
Liyi, I don't think you should do this.  Instead, let us try to understand the instances in the C semantics where you really need this feature.  Can you send us one or two of them, as well as your assessment of what it takes to change the code to avoid it?

Grigore

**Comment by grosu** [comment]
In my view, ending up with such configurations:

```
<k>
    checkResult ( HOLE ) ~> maybeSymVal ( tv ( 1 , 1 ) , HOLE )
</k>
```

is terribly wrong.  It shows a poor semantic definitional style, which should be avoided.  You should not end up with a context at top of the k cell.

**Comment by grosu** [comment]
So are you saying that this is a front-end issue?

Grigore

**Comment by yilongli** [comment]
I am not sure. If you write `rule checkResult(HOLE) => noResult`, the term `checkResult(HOLE)` will be created as a normal KItem by the front-end instead of a freezer. This is quite an unusual way of writing rule, so I am not sure what the expected behavior should be.

**Comment by dwightguth** [comment]
I believe that the syntax for writing freezers manually in rules is still:

Syntax KLabel ::= #freezer(K)

Check the prelude.

**Comment by liyili2** [comment]
Actually, I eventually came up with a solution for this problem. My
solution is that to add a intermediate step to produce the K Sequence.

```
module TEST

     syntax KItem ::= maybeSymVal(K, K) [strict]
                | checkResult(K) | "start" | "abc"

     syntax KResult ::= "noResult" | "symVal" | TypedValue

     rule checkResult(K)  => K ~> abc

     rule R:KResult ~> abc => R
     rule abc => noResult

     syntax TypedValue ::= tv(K, Int)

     syntax PureStatement ::= "BlockStatement" "(" K ")"
          [klabel('BlockStatement)]

     rule start =>
maybeSymVal(checkResult(BlockStatement(1)),checkResult(BlockStatement(0)))

     rule BlockStatement(1) => tv(1,1)
     rule BlockStatement(0) => .K

     rule maybeSymVal(_:KResult, noResult) => .K
     rule maybeSymVal(tv(_, T:Int), tv(_, T:Int)) => tv(symVal, T)

endmodule
```

**Comment by grosu** [comment]
This is cleaner anyway.  You may even want to do what I did in the tutorial in some places, namely something like this:

```
rule start => check(arg1,arg2) ~> maybeOp(arg1,arg2)
```

Grigore

**Comment by grosu** [comment]
But do we really need freezers anymore?  Can't they just be KItems?  We definitely needed them when we allowed heating/cooling to take place anywhere, but now that we only allow heating/cooling at the top of the k cell ...

Traian?

**Comment by traiansf** [comment]
maybe we don't, but are we committed to the fact that we will _never_ consider heating/cooling outside of the top of the <k> cell? I'm not sure we made a theoretical commitment to that.
one additional  things about freezers is that if we don't freeze them, they need to be in a _strong normal form_, i.e. no rules (either structural or computational) are allowed to apply on the pattern; otherwise we might not be able to cool.

Now we "solve" this theoretically by saying that we create a new label representing the frozen pattern.

**Comment by grosu** [comment]
The clean solution is to keep the freezers, of course, but then we need to find a clean way to deal with them at the KAST level.  If a freezer is indeed just one KLabel and that is consistently used as such whenever a HOLE appears, then everything seems to work fine.  But note what Yilong said:

> If you write rule checkResult(HOLE) => noResult, the term checkResult(HOLE)
> will be created as a normal KItem by the front-end instead of a freezer. 

So what is the freezer then?  I think it should a freezer, which should also be a KItem.  Either a KToken `#freezerCheckResult`(HOLE`)` or a KLabel applied to an emply KList `#freezerCheckResult`(HOLE`)(.::KList)`.

Maybe we should consider this issue a front-end issue, and have @radumereuta fix it.

I'm going to start a new issue on freezers.

**Comment by radumereuta** [comment]
I think it would be wrong to consider this a front end problem.
I'm trying to make the front end use as few classes as possible (remove bracket, remove special classes for list terminator, remove special representation for builtin tokens). I want to have only two things appear in the AST after the parser: TermCons and Constant (for generic lexical productions).

The way I see it should be fixed, is to transform the TermCons/KLabel into a freezer class (or whatever it's needed for the backend) after the front end finishes parsing. I think the best place to do it, is in FlattenSyntax.

**Comment by grosu** [comment]
The freezer is just a KLabels, isn't it?  The fact that you parse HOLE as a K term is actually wrong sematically, because there is no HOLE in fact, it is only syntactic sugar for a name.  We may enlarge the scope of the front-end, but from an abstract syntax or  semantic perspective, there is no HOLE, there are only some special KLabels that for some reason we decided to call freezers.

Grigore

**Comment by traiansf** [comment]
It might be nicer (at least in kast) to represent freezers as terms as we
can do perform more analysis on them this way.

Nevertheless, once in the backend they should be nothing more than labels
(although I must confess now that in the Maude backend I preferred keeping
them as terms.

One thing to keep in mind is that we might want to have freezers which are
more complex such as the ones used for the call-by-need lambda calculus,
where one heats only if the current k sequence is in a certain form, and
freezes an entire K sequence in the process.

2014-07-25 19:48 GMT+03:00 Grigore Rosu notifications@github.com:
